pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/StablePathMatcher.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.rule.internal;
+
+import java.util.ArrayDeque;
+import java.util.Objects;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.lang.java.ast.ASTExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTFieldAccess;
+import net.sourceforge.pmd.lang.java.ast.ASTMethodCall;
+import net.sourceforge.pmd.lang.java.ast.ASTSuperExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTThisExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTVariableAccess;
+import net.sourceforge.pmd.lang.java.symbols.JVariableSymbol;
+
+/**
+ * A matcher for an expression like {@code a}, {@code a.b}, {@code a.getFoo()}.
+ * Those are expressions we assume to be pure, and to be referring to
+ * the same reference when they're called repeatedly if no side effect
+ * occurs between calls.
+ *
+ * <p>Note that this is not relocatable: you must use a matcher in the
+ * same scope it has been created in, to avoid bugs with accessibility/shadowing/etc.
+ * You must take care yourself that no side-effect occurs.
+ */
+public final class StablePathMatcher {
```

Comment:
Note: this thing, might look overweight for the task (there's a single test case in BrokenNullCheck that requires it), but I think it could be used in other rules to make them smarter. For instance, in ForLoopCanBeForeach, we could recognise looping on any "stable path", eg `this.field.collection` or `this.getCollection()`.